data-structures/LinkedList.py

```diff
@@ -1,121 +1,110 @@
+        current_item = self.head
+        while current_item.next.next:          #Find second to last element
+            current_item = current_item.next
 
-        value = current.next.value
-        current.next = None
+        value = current_item.value
+        current_item.next = None
         self.total -= 1
         return value
 
-    def get_at_index(self, i):
-        """ return element at index i """
+    def get_value(self, i):
+        """Return element at the index i"""
         if i >= self.total or i < 0:
             raise IndexError
 
+        current_item = self.head
         current_index = 0
-        current = self.head
         while current_index != i:
-            current = current.next
+            current_item = current_item.next
             current_index += 1
+        return current_item.value
 
-        return current.value
-
-    def insert_at_index(self, i, value):
-        """ insert an element at a certain index i """
-
+    def insert(self, i, value):
+        """Insert an element at index i"""
         if i > self.total or i < 0:
             raise IndexError
 
-        if i == self.total:
-            self.add(value)
-            return
-
+        current_item = self.head
         current_index = 0
-        current = self.head
-        while current_index != i - 1:  # make current is element before i
-            current = current.next
+        while current_index != i - 1:
+            current_item = current_item.next
             current_index += 1
 
-        # current = element at index i - 1
-        # current.next = element at index i (we are going to move it to i + 1)
-        new_node = Node(value)
-        new_node.next = current.next  # what ever is at index i now will be i + 1
-        current.next = new_node  # make i-1.next  = new node
+        new_item = Node(value)
+        new_item.next = current_item.next
+        current_item.next = new_item
         self.total += 1
 
     def remove_at_index(self, i):
-        """ remove an element at index i """
+        """Remove element at index i"""
         if i >= self.total or i < 0:
             raise IndexError
 
-        if i == self.total:
-            self.pop()
-            return
-
+        current_item = self.head
         current_index = 0
-        current = self.head
-        while current_index != i - 1:  # make current is element before i
-            current = current.next
+        while current_index != i - 1:
+            current_item = current_item.next
             current_index += 1
 
-        node_at_i_plus_1 = current.next.next  # element at i + 1
-        current.next = node_at_i_plus_1
+        current_item.next = current_item.next.next
+        self.total -= 1
+
+    def remove_value(self, value):
+        """Remove the first element that contains value"""
+        current_item = self.head
+        while current_item.next.value != value:  # Find the element before the element that contains said value
```

Comment:
this will break as soon as you hit the end of the list. tail.next=None, so None.value will through an error. you have to check that it exists before you access it.